bentoml/server/bento_api_server.py

```diff
@@ -106,16 +116,18 @@ def start(self):
         # Flask dev server enabled threaded by default, disable it.
         self.app.run(port=self.port, threaded=False)
 
-    @staticmethod
-    def index_view_func():
+    def index_view_func(self):
         """
         The index route for BentoML API server
         """
-        return Response(
-            response=INDEX_HTML.format(url='/docs.json'),
-            status=200,
-            mimetype="text/html",
-        )
+        if hasattr(self.bento_service, '_static_files'):
+            return render_template('index.html')
```

Comment:
In case custom UI is enabled, do we want to also have `swagger` UI on some different endpoint?

Reply:
Yea, that'd be great for debugging, I think we can add a `/swagger.html` endpoint which always renders the swagger UI

Reply:
Does it actually need the `redner_template` method? the user can not really use the templating functionality right?  I'm more inclined to simply put all the static files under the `static` directory if they are just static files

Reply:
Makes sense!

so now the dir structure will be:
```
webui_dir/
     |-- js/
     |--css/
     |--imgs/
     |--index.html
```

instead of `render_template` now it uses `send_from_directory`

Reply:
> Yea, that'd be great for debugging, I think we can add a `/swagger.html` endpoint which always renders the swagger UI

In my opinion, `/docs` endpoint would be better for `swagger UI` since this is the same as what `fastapi` uses. And also if in future we change from `swagger` to something else, the path can remain the same. 

So `/docs` for the HTML page and `/docs.json` for openapi json.

What do you think?

Reply:
@Korusuke Yea I think `/docs` for the UI page and `/docs.json`for the json sounds great!

Reply:
What's the reason for `send_from_directory` instead of just the `static_folder` approach?

Reply:
> What's the reason for `send_from_directory` instead of just the `static_folder` parameter?

For serving `index.html` on `/` endpoint. if we don't use `send_from_directory` then only `/index.html` will work

Reply:
I believe it should serve all files under the static_folder directory not just index.html, you may also need to specify static_url_path: `app = Flask(__name__, static_folder='../build', static_url_path='/')`
and in your index view function do something like ```app.send_static_file('index.html')```

Reply:
I think this way it is more friendly to the web frontend developer, for example, if the user creates the frontend project with something like https://github.com/facebook/create-react-app or use webpack to build their js project, they can simply supply the built `dist` directory for BentoML to serve as static files, and everything should just work.

Reply:
> I believe it should serve all files under the static_folder directory not just index.html, you may also need to specify static_url_path: `app = Flask(__name__, static_folder='../web_client', static_url_path='/')`
> and in your index view function do something like `app.send_static_file('index.html')`

yeh, this is working currently!

> Adding static_url_path might break the swagger js/css files tho, you may need to adjust accordingly

Fixed that using send_from_directory and changing swagger static dir name.

> I think this way it is more friendly to the web frontend developer, for example, if the user creates the frontend project with something like https://github.com/facebook/create-react-app or use webpack to build their js project, they can simply supply the built `dist` directory for BentoML to serve as static files, and everything should just work.

Yep, I am trying to achieve this...currently the only issue is serving `any_path/index.html` for  `any_path/`, `any_path/index.html` and also `any_path/index`.

trying to find some better solution instead of just writing down all the cases.

Reply:
One another doubt is - should we allow user-defined routes to override the system defined (predefined by bento)?

currently, if the user-defined function name is `index` then the server doesn't start, so should we allow to override that?
this may provide some extra flexibility but may even be a bit confusing at times.

Reply:
@Korusuke  - I think any of the `@api` routes and bentoml routes such as `/healthz` and `/metrics` should take place before the static folder.  I thought that's already the case for the `static_folder` parameter with Flask?

> currently, if the user-defined function name is index then the server doesn't start, so should we allow to override that?

`.. the server doesn't start ..`  is that because it has added two rules handling the same path? Maybe we should validate the `@api` name and make sure it is not in the reserved name list, such as `index`, `healthz`, `metrics` etc.  Currently there's only a `isidentifier` check, you can add this validation here: https://github.com/bentoml/BentoML/blob/v0.8.2/bentoml/service.py#L311

Reply:
Okey great!
should I include this in the same PR or create another issue/PR for this?

Reply:
@Korusuke a smaller PR that adds the validation would be nice, but it's ok to just include it in this PR as well, whichever is easier for you